### ISSUE TITLE
Re-enable skipped blackhole distributed tests

### ISSUE THREAD

**Issue by williamlyTT** [tenstorrent/tt-metal#19751]
Blackhole post commit currently failing on test:
```
[ RUN      ] MeshEventsTestSuite.ReplicatedAsyncIO
2025-03-17 17:49:22.399 | INFO     | SiliconDriver   - Opened PCI device 0; KMD version: 1.31.0, IOMMU: disabled
                  Metal | INFO     | Initializing device 0. Program cache is NOT enabled
                 Device | INFO     | For Blackhole hardcode AICLK to 1350 MHz due to lack of ARC message support
                  Metal | INFO     | AI CLK for device 0 is:   1350 MHz
libc++abi: terminating due to uncaught exception of type std::runtime_error: No core coordinate found at location: (0, 0, ETH, LOGICAL)
```

Affected Test(s) to be skipped for blackhole:
- MeshEventsTestSuite.ReplicatedAsyncIO
- MeshEventsTestSuite.AsyncWorkloadAndIO
- MeshEventsTestSuite.CustomDeviceRanges
- MeshEventsTestSuite.MultiCQNonBlockingReads

Fix for the test crash is in latest UMD, but blocked because bumping UMD causes issues in TG that require a FW fix.

**Comment by williamlyTT** [comment]
cc @omilyutin-tt @abhullar-tt @broskoTT @akirby-TT

**Comment by williamlyTT** [comment]
PR merged https://github.com/tenstorrent/tt-metal/pull/19761 
Please revert the workflow change to test in CI

**Comment by broskoTT** [comment]
Just to update, this issue is not UMD related, I'll copy my message from slack:
FYI
@cfjchu @aliuTT 
We were trying to figure out the issue with the failing BH test, and here's the culprit:
This line defines which dispatch cores are used in the test: https://github.com/tenstorrent/tt-metal/blob/2ad6823ae35af754520e70c76b5771f762f0e110/tests/tt_metal/tt_metal/common/multi_device_fixture.hpp#L144
The resulting core descriptor would be this one: https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/core_descriptors/blackhole_140_arch_eth_dispatch.yaml
It tries to use ethernet cores for dispatch. On the P100 cards we don't have any ethernet cores available, all are harvested, and hence the failure.
How do you propose to address this? We could maybe test from MeshDeviceType if the card is P100, but I see that P100 is not in the list of this enum at all? And arch + number of chips is not enough to differentiate between p100 and p150 ...

**Comment by cfjchu** [comment]
We should have a separate enum for differentiating between P100 and P150 and implement what you suggest.

As a quick fix, could we just query the number of ethernet cores available and use that information to do the select?

**Comment by broskoTT** [comment]
@cfjchu not sure how do I extract this at this level. Can you take the fix?

**Comment by bbradelTT** [comment]
I'm getting this same error when I run the following on a p300a (yyzo-bh-02)
```
import torch
import ttnn

device_id = 0
device = ttnn.open_device(device_id=device_id)
print("Opened Device")
ttnn.close_device(device)
print("Closed Device")
```
Output:
```
python open_device.py 
2025-03-28 11:53:52.671 | DEBUG    | ttnn:<module>:83 - Initial ttnn.CONFIG:
Config{cache_path=/home/bbradel/.cache/ttnn,model_cache_path=/home/bbradel/.cache/ttnn/models,tmp_dir=/tmp/ttnn,enable_model_cache=false,enable_fast_runtime_mode=true,throw_exception_on_fallback=false,enable_logging=false,enable_graph_report=false,enable_detailed_buffer_report=false,enable_detailed_tensor_report=false,enable_comparison_mode=false,comparison_mode_should_raise_exception=false,comparison_mode_pcc=0.9999,root_report_path=generated/ttnn/reports,report_name=std::nullopt,std::nullopt}
2025-03-28 11:53:52.720 | INFO     | SiliconDriver   - Opened PCI device 0; KMD version: 1.29.0, IOMMU: disabled
2025-03-28 11:53:52.802 | INFO     | SiliconDriver   - Opened PCI device 1; KMD version: 1.29.0, IOMMU: disabled
2025-03-28 11:53:53.009 | INFO     | SiliconDriver   - Opened PCI device 0; KMD version: 1.29.0, IOMMU: disabled
2025-03-28 11:53:53.095 | INFO     | SiliconDriver   - Opened PCI device 1; KMD version: 1.29.0, IOMMU: disabled
2025-03-28 11:53:53.309 | INFO     | SiliconDriver   - Harvesting mask for chip 0 is 0x88 (physical layout: 0x2040, logical: 0x88, simulated harvesting mask: 0x0).
2025-03-28 11:53:53.316 | INFO     | SiliconDriver   - Opened PCI device 0; KMD version: 1.29.0, IOMMU: disabled
2025-03-28 11:53:53.395 | INFO     | SiliconDriver   - Harvesting mask for chip 1 is 0x30 (physical layout: 0x500, logical: 0x30, simulated harvesting mask: 0x0).
2025-03-28 11:53:53.403 | INFO     | SiliconDriver   - Opened PCI device 1; KMD version: 1.29.0, IOMMU: disabled
2025-03-28 11:53:53.479 | INFO     | SiliconDriver   - Detected PCI devices: [0, 1]
2025-03-28 11:53:53.479 | INFO     | SiliconDriver   - Using local chip ids: {0, 1} and remote chip ids {}
2025-03-28 11:53:53.962 | INFO     | SiliconDriver   - Device: 0 Mapped iATU region 0 from 0x0 to 0x3fffffff to 0x2c0000000
2025-03-28 11:53:53.962 | INFO     | SiliconDriver   - Device: 0 Mapped iATU region 1 from 0x40000000 to 0x7fffffff to 0x280000000
2025-03-28 11:53:53.962 | INFO     | SiliconDriver   - Device: 1 Mapped iATU region 0 from 0x0 to 0x3fffffff to 0x1c0000000
2025-03-28 11:53:53.962 | INFO     | SiliconDriver   - Device: 1 Mapped iATU region 1 from 0x40000000 to 0x7fffffff to 0x180000000
libc++abi: terminating due to uncaught exception of type std::out_of_range: map::at:  key not found
Aborted (core dumped)
```

I pulled and synced submodules a few minutes ago.

How do I get around this?

**Comment by bbradelTT** [comment]
I seem to still be running into this on test dispatch at the umd bump commit https://github.com/tenstorrent/tt-metal/actions/runs/14128270495/job/39582480726

**Comment by xanderchin** [comment]
> [tt-metal/tests/tt_metal/tt_metal/common/multi_device_fixture.hpp](https://github.com/tenstorrent/tt-metal/blob/2ad6823ae35af754520e70c76b5771f762f0e110/tests/tt_metal/tt_metal/common/multi_device_fixture.hpp#L144)
> 
> Line 144 in [2ad6823](/tenstorrent/tt-metal/commit/2ad6823ae35af754520e70c76b5771f762f0e110)
>  auto core_type = (config_.num_cqs >= 2 and *mesh_device_type != MeshDeviceType::TG) ? DispatchCoreType::ETH 
> 
> The resulting core descriptor would be this one: https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/core_descriptors/blackhole_140_arch_eth_dispatch.yaml
> It tries to use ethernet cores for dispatch. On the P100 cards we don't have any ethernet cores available, all are harvested, and hence the failure.
> How do you propose to address this? We could maybe test from MeshDeviceType if the card is P100, but I see that P100 is not in the list of this enum at all? And arch + number of chips is not enough to differentiate between p100 and p150 ...

@cfjchu just bubbling this up again. Is anyone looking into this issue? It's also currently assigned to @broskoTT , but it appears to not be a UMD issue.

**Comment by gfengTT** [comment]
FWIW I've noticed the same error in other places where `MeshDevice::create()` is called.

**Comment by cfjchu** [comment]
Yeah, let me take a look today. I think it's related to tests doing dispatch on row which isn't supported on BH

**Comment by gfengTT** [comment]
@cfjchu I did some investigation and the error seems to be occurring when calling `virtual_core_from_logical_core` in `Device::initialize_allocator()`.

On p100, all of `(0,0)`..`(0,13)` are logical dispatch cores, and the error occurs on `(0,0)`: `No core coordinate found at location: (0, 0, ETH, LOGICAL)`.

On p150, 0,1,2,3,6,12,13 are logical dispatch cores, and 4,5,7,8,9,10,11 are active ethernet cores. The error occurs on `(0,12)`: `No core coordinate found at location: (0, 12, ETH, LOGICAL)`. The conversion on 0,1,2,3,6 all passed.

**Comment by gfengTT** [comment]
Some more digging. It happens here: 

https://github.com/tenstorrent/tt-metal/blob/e0e0d821fc949d0d9d10d3d7681b66bf893c9342/tt_metal/llrt/tt_cluster.cpp#L472

where coord is (0,12), type is ETH. @broskoTT hope that helps.

**Comment by cfjchu** [comment]
Fix and re-enabling the tests are here: https://github.com/tenstorrent/tt-metal/pull/20558

**Comment by broskoTT** [comment]
@gfengTT our p100 boards don't have any functional eth cores, so tt-metal shouldn't request them.
for p150, I assume you've hit a board with harvested eth cores, in that case there are only 12 active eth cores (up to 11) compared to full number of eth cores of 14
Just wanted to clarify. But @cfjchu 's fix is fine, it will use worker cores for dispatch for blackhole

**Comment by gfengTT** [comment]
@broskoTT right. It does have 2 harvested channels (harvesting mask 576).
@cfjchu thanks for the fix! Should we add a check for harvesting somewhere?
